## How SQL Work - Understanding the implementation
![Imgur](https://i.imgur.com/g55maqS.png)

# &#128209; Contents
<ul>
<li>1. Introduction</li>
<ul>
    <li>1.1 Defination Of SQL </li>
    <li>1.2 Relational Databases</li>
    <li>1.3 Connection Database </li>
</ul>
<li>2. Implementation </li>
<ul>
    <li>2.1 Import Libraries </li>
    <li>2.2 Database connection code </li>
    <li>2.3  Initial exploration of the Dataset </li>
    <li>2.4 Overview of Analysis </li>
    <li>2.5 SQL Code Implementation</li>
</ul>
</ul>

<br>

# &#128253; 1. Introduction

In [1]:
%%html
<style>
@import url('https://fonts.googleapis.com/css?family=Ewert|Roboto&effect=3d|ice|');
body {background-color: gainsboro;} 
a {color: #37c9e1; font-family: 'Roboto';} 
h1 {color: #37c9e1; font-family: 'Orbitron'; text-shadow: 4px 4px 4px #aaa;} 
h2, h3 {color: slategray; font-family: 'Orbitron'; text-shadow: 4px 4px 4px #aaa;}
h4 {color: #818286; font-family: 'Roboto';}
span {font-family:'Roboto'; color:black; text-shadow: 5px 5px 5px #aaa;}  
div.output_area pre{font-family:'Roboto'; font-size:110%; color:lightblue;}      
</style

## &#128203; 1.1  Defination of SQL:
SQL is a conceptual language for working with data stored in databases. In our case, SQLite is the specific implementation. Most SQL languges share all of the capabilities in this doc. The differences are usually in performance and advances analytical funcionalities (and sometimes bugs of course). Eventually, we will use SQL lunguage to write queries that would pull data from the DB, manipulate it, sort it, and extract it.

![Imgur](https://i.imgur.com/3V5m8s5.jpg)

The most important component of the DB is its tables - that's where all the data stored. Usually the data would be devided to many tables, and not stored all in one place (so designing the data stracture propely is very important). Most of this script would handle how to work with tables. Other than tables, there are some other very useful concepts/features that we won't talk about:

* table creation

* inserting / updating data in the DB

* functions - gets a value as an input, and returns manipulation of that value (for example function that remove white spaces)

## &#128206; 1.2  Relational Databases:
A relational database is, simply, a database that stores related information across multiple tables and allows you to query information in more than one table at the same time.

It's easier to understand how this works by thinking through an example. Imagine you're a business and you want to keep track of your sales information. You could set up a spreadsheet in Excel with all of the information you want to keep track of as separate columns: Order number, date, amount due, shipment tracking number, customer name, customer address, and customer phone number.

![Imgur](https://i.imgur.com/QhbmFgq.png)
 
 Sales data in Excel &#128070;
 
![Imgur](https://i.imgur.com/KCj0fP8.png)

Sales data in a relational database &#128070;

With a relational database, like the one in the above diagram, you avoid all of these issues. You could set up two tables, one for orders and one for customers. The 'customers' table would include a unique ID number for each customer, along with the name, address and phone number we were already tracking. The 'orders' table would include your order number, date, amount due, tracking number and, instead of a separate field for each item of customer data, it would have a column for the customer ID. This enables us to pull up all of the customer info for any given order, but we only have to store it once in our database rather than listing it out again for every single order.

## &#128268; 1.3  Connection Database:
The basic structure of the query is very simple: You define what you want to see after the SELECT, * means all possible columns You choose the table after the FROM You add the conditions for the data you want to use from the table(s) after the WHERE
![Imgur](https://i.imgur.com/gSEqsbo.jpg)
The stracture, and the order of the sections matter, while spaces, new lines, capital words and indentation are there to make the code easier to read.

# &#128187; Implementation

## &#128229; 2.1  Import Libraries
We will start by importing the Python libraries we will be using in this analysis. These libraries include:

* sqllite3 for interacting with a local relational database; 

* pandas  for data ingestion .

In [2]:
import sqlite3
import pandas as pd

## &#128268; 2.2 Database Connection Code

In [3]:
path = "../input/" 
database = path + 'hubway.db'

conn = sqlite3.connect(database)
tables = pd.read_sql("""SELECT *
                        FROM sqlite_master
                        WHERE type='table';""", conn)
print("Conection SuccessFull",conn)

Conection SuccessFull <sqlite3.Connection object at 0x7f51d66f3570>


## &#128197; 2.3 Initial exploration of the Dataset
Let's start by taking a look at our database. The database has two tables, trips and stations. To begin with, we'll just look at the trips table, which contains the following columns:

*` id `— A unique integer that serves as a reference for each trip.

*`duration` — The duration of the trip, measured in seconds.

*`start_date` — The date and time the trip began.

*`start_station` — An integer that corresponds to the id column in the stations table for the station the trip started.

*`end_date` — The date and time the trip ended.

*` end_station` — The 'id' of the station the trip ended.

*` bike_number` — Hubway's unique identifier for the bike used on the trip.

*`sub_type` — The subscription type of the user. "Registered" for users with a membership, "Casual" for users without a membership.

*`zip_code`— The zip code of the user (only available for registered members).

*`birth_date` — The birth year of the user (only available for registered members).

*`gender` — The gender of the user (only available for registered members).

Now, we will need to read the dataset using the commands below.

Note: Make sure Database connection before you run the data load and read code below.

df is a variable pointing to a pandas data frame.

In [4]:
df = pd.read_sql_query("SELECT * FROM trips", conn)
df

id  duration    ...    birth_date    gender
0              1         9    ...          1976      Male
1              2       220    ...          1966      Male
2              3        56    ...          1943      Male
3              4        64    ...          1981    Female
4              5        12    ...          1983    Female
5              6        19    ...          1951      Male
6              7        24    ...          1971    Female
7              8         7    ...          1971    Female
8              9         8    ...          1983    Female
9             10      1108    ...          1994      Male
10            11      1055    ...          1956      Male
11            12      1042    ...          1959    Female
12            13       994    ...                        
13            14        15    ...          1982      Male
14            15       952    ...          1944      Male
15            16        32    ...          1960    Female
16            17      1261    ...          1975      Male
17            18         7    ...          1956      Male
18            19      1020    ...          1987    Female
19            20      1264    ...          1985    Female
20            21      1043    ...          1978      Male
21            22       969    ...          1971      Male
22            23      1582    ...          1980    Female
23            24       620    ...          1955    Female
24            25         7    ...          1971    Female
25            26       771    ...          1943      Male
26            27      1222    ...          1972    Female
27            28         4    ...          1949      Male
28            29         5    ...          1968      Male
29            30       824    ...          1967      Male
...          ...       ...    ...           ...       ...
1569971  1578995       360    ...                    Male
1569972  1578996       540    ...                    Male
1569973  1578997       300    ...                    Male
1569974  1578998       420    ...                    Male
1569975  1579000       300    ...                        
1569976  1579001      1200    ...                    Male
1569977  1579002       360    ...                    Male
1569978  1579003       300    ...                    Male
1569979  1579004       300    ...                    Male
1569980  1579005       960    ...                    Male
1569981  1579006       180    ...                  Female
1569982  1579007       420    ...                  Female
1569983  1579008       300    ...                    Male
1569984  1579009       300    ...                    Male
1569985  1579010       780    ...                    Male
1569986  1579011       540    ...                    Male
1569987  1579012      1440    ...                    Male
1569988  1579013       480    ...                    Male
1569989  1579014       420    ...                  Female
1569990  1579015       360    ...                    Male
1569991  1579016       300    ...                  Female
1569992  1579017       120    ...                    Male
1569993  1579018       900    ...                  Female
1569994  1579019       240    ...                    Male
1569995  1579020       360    ...                    Male
1569996  1579021       720    ...                    Male
1569997  1579022       480    ...                    Male
1569998  1579023       540    ...                        
1569999  1579024       120    ...                    Male
1570000  1579025       420    ...                Female\n

[1570001 rows x 11 columns]

## &#128240; 2.4 Overview of Analysis
In this Notebook I will  try cover up some Question By SQL Code or Command,Here are some Question:

* What was the duration of the longest trip?

* How many trips were taken by 'registered' users?

* What was the average trip duration?

* Do registered or casual users take longer trips?

* Which bike was used for the most trips?

* What is the average duration of trips by users over the age of 30?

## &#128187; 2.5 SQL Code Implementation
The SQL code or commands we'll use to answer these questions are:
<ul>
    <li>1. SELECT</li>
    <li>2. LIMIT</li>
    <li>3.  ORDER BY</li>
    <li>4.WHERE</li>
    <li>5.COUNT</li>
    <li>6.Aggregate Functions(SUM,AVG,MIN,MAX)</li>
    <li>7.GROUP BY</li>
    <li>8. Arithmetic Operators</li>
    <li>9. JOIN</li>
    </ul>

# &#128505; 1. SELECT 
SELECT will be the foundation of almost every query you write - it tells the database which columns you want to see. You can either specify columns by name (separated by commas) or use the wildcard `*` to return every column in the table. In addition to the columns you want to retrieve, you also have to tell the database which table to get them from. To do this you use the keyword `FROM` followed by the name of the table. For example, if we wanted to see the start_date and bike_number for every trip in the trips table, we could use the following query:

In [5]:
pd.read_sql_query("SELECT start_date, bike_number FROM trips;",conn)

start_date bike_number
0        2011-07-28 10:12:00      B00468
1        2011-07-28 10:21:00      B00554
2        2011-07-28 10:33:00      B00456
3        2011-07-28 10:35:00      B00554
4        2011-07-28 10:37:00      B00554
5        2011-07-28 10:39:00      B00456
6        2011-07-28 10:47:00      B00554
7        2011-07-28 10:48:00      B00554
8        2011-07-28 11:01:00      B00554
9        2011-07-28 11:55:00      B00550
10       2011-07-28 11:55:00      B00580
11       2011-07-28 11:55:00      B00539
12       2011-07-28 12:00:00      B00368
13       2011-07-28 12:00:00      B00442
14       2011-07-28 12:00:00      B00556
15       2011-07-28 12:00:00      B00316
16       2011-07-28 12:00:00      B00454
17       2011-07-28 12:00:00      B00404
18       2011-07-28 12:01:00      B00147
19       2011-07-28 12:01:00      B00165
20       2011-07-28 12:01:00      B00527
21       2011-07-28 12:01:00      B00487
22       2011-07-28 12:01:00      B00093
23       2011-07-28 12:01:00      B00375
24       2011-07-28 12:02:00      B00275
25       2011-07-28 12:02:00      B00279
26       2011-07-28 12:02:00      B00074
27       2011-07-28 12:02:00      B00137
28       2011-07-28 12:02:00      B00082
29       2011-07-28 12:02:00      B00460
...                      ...         ...
1569971  2013-11-30 20:17:00      T01149
1569972  2013-11-30 20:17:00      T01268
1569973  2013-11-30 20:23:00      T01352
1569974  2013-11-30 20:23:00      T01185
1569975  2013-11-30 20:33:00      T01217
1569976  2013-11-30 20:46:00      T01146
1569977  2013-11-30 20:50:00      T01328
1569978  2013-11-30 20:59:00      T01182
1569979  2013-11-30 21:01:00      B01468
1569980  2013-11-30 21:07:00      T01439
1569981  2013-11-30 21:10:00      T01351
1569982  2013-11-30 21:19:00      T01175
1569983  2013-11-30 21:41:00      T01343
1569984  2013-11-30 21:47:00      B01489
1569985  2013-11-30 21:54:00      T01350
1569986  2013-11-30 21:55:00      T01175
1569987  2013-11-30 22:15:00      B01480
1569988  2013-11-30 22:25:00      T01308
1569989  2013-11-30 22:25:00      T01043
1569990  2013-11-30 22:51:00      T01352
1569991  2013-11-30 22:53:00      T01217
1569992  2013-11-30 23:07:00      B01468
1569993  2013-11-30 23:17:00      B00585
1569994  2013-11-30 23:24:00      T01195
1569995  2013-11-30 23:24:00      T01069
1569996  2013-11-30 23:30:00      T01341
1569997  2013-11-30 23:32:00      T01328
1569998  2013-11-30 23:32:00      T01310
1569999  2013-11-30 23:36:00      B01478
1570000  2013-11-30 23:39:00      T01352

[1570001 rows x 2 columns]

### &#128681; Note:
In this example, we started with the SELECT command so that the database knows we want it to find us some data. Then we told the database we were interested in the start_date and bike_number columns. Finally we used FROM to let the database know that the columns we want to see are part of the trips table.

One important thing to be aware of when writing SQL queries is that the language requires every query to end with a semicolon (;).

# &#128208; 2. LIMIT 
The next command we need to know before we start to run queries on our Hubway database is LIMIT. LIMIT simply tells the database how many rows you want it to return. The SELECT query we looked at in the previous section would return the requested information for every row in the trips table, but sometimes that could mean a lot of data and you might not want all of it. If, instead, we wanted to see the start_date and bike_number for the first five trips in the database, we could add LIMIT to our query as follows:

In [6]:
pd.read_sql_query("SELECT start_date, bike_number FROM trips LIMIT 5;",conn)

start_date bike_number
0  2011-07-28 10:12:00      B00468
1  2011-07-28 10:21:00      B00554
2  2011-07-28 10:33:00      B00456
3  2011-07-28 10:35:00      B00554
4  2011-07-28 10:37:00      B00554

We simply added the LIMIT command and then a number representing the number of rows we want to be returned. In this instance we used 5, but you can replace that with any number to get the appropriate amount of data for the project you're working on.

We will use LIMIT a lot in our queries on the Hubway database in this notebook - the trips table contains over a 1.5 million rows of data and we certainly don't need to display all of those to answer the questions we're going to be working on.

Let's run our first query on the Hubway database. First we will store our query as a string and then use the function we defined earlier to run it on the database. Take a look at the following example:

In [7]:
df =pd.read_sql_query("SELECT * FROM trips LIMIT 5;",conn) 
df

id  duration           start_date   ...    zip_code birth_date  gender
0   1         9  2011-07-28 10:12:00   ...      '97217     1976.0    Male
1   2       220  2011-07-28 10:21:00   ...      '02215     1966.0    Male
2   3        56  2011-07-28 10:33:00   ...      '02108     1943.0    Male
3   4        64  2011-07-28 10:35:00   ...      '02116     1981.0  Female
4   5        12  2011-07-28 10:37:00   ...      '97214     1983.0  Female

[5 rows x 11 columns]

This query uses`*` as a wildcard instead of specifying columns to return. This means the `SELECT` command has given us every column in the trips table. We also used the`LIMIT` function to restrict the output to the first five rows of the table.

The previous example returned every column in the trips table. If we were only interested in the duration and start_date columns, we could replace the wildcard with the column names as follows:

In [8]:
df=pd.read_sql_query("SELECT duration, start_date FROM trips LIMIT 5",conn)
df

duration           start_date
0         9  2011-07-28 10:12:00
1       220  2011-07-28 10:21:00
2        56  2011-07-28 10:33:00
3        64  2011-07-28 10:35:00
4        12  2011-07-28 10:37:00

# &#128225; 3. ORDER BY
The final command we need to know before we can answer the first of our questions is `ORDER BY`. This command allows you to sort the database on a given column. To use it, you simply specify the name of the column you would like to sort on. By default, `ORDER BY` sorts in ascending order. If you would like to specify which order the database should be sorted, you can add the keyword `ASC` for ascending order or `DESC` for descending order. For example, if we wanted to sort the trips table from the shortest duration to the longest we could add the following line to our query:

In [9]:
df=pd.read_sql_query("SELECT duration FROM trips ORDER BY duration ASC",conn)
df

duration
0               0
1               0
2               0
3               0
4               0
5               0
6               0
7               0
8               0
9               0
10              0
11              0
12              0
13              0
14              0
15              0
16              0
17              0
18              0
19              0
20              0
21              0
22              0
23              0
24              0
25              0
26              0
27              0
28              0
29              0
...           ...
1569971      9974
1569972      9974
1569973      9976
1569974      9978
1569975      9980
1569976      9980
1569977      9981
1569978      9982
1569979      9984
1569980      9985
1569981      9985
1569982      9986
1569983      9986
1569984      9987
1569985      9988
1569986      9989
1569987      9991
1569988      9992
1569989      9993
1569990      9994
1569991      9994
1569992      9994
1569993      9995
1569994      9995
1569995      9996
1569996      9996
1569997      9997
1569998      9998
1569999      9998
1570000      9999

[1570001 rows x 1 columns]

**3.1** With the `SELECT`, `LIMIT` and `ORDER BY` commands in our repertoire, we can now attempt to answer our first question: **What was the duration of the longest trip?**

To answer this question, it's helpful to break it down into sections and identify which commands we will need to address each part. First we need to pull the information from the duration column of the trips table. Then, to find which trip is the longest, we can sort the duration column in descending order. Here's how you might work this through to come up with a query that will get the information you're looking for:

* Use `SELECT` to retrieve the duration column FROM the trips table.

* Use `ORDER BY` to sort the duration column and use the `DESC` keyword to specify that you want to sort in descending order.

* Use `LIMIT` to restrict the output to 1 row.

Let's go ahead and run this query and find out how long the longest trip lasted.

In [10]:
df=pd.read_sql_query("SELECT duration FROM trips ORDER BY duration DESC LIMIT 1;",conn)
df

duration
0      9999

Now we know that the longest trip lasted 9999 seconds, or a little over 166 minutes. With a maximum value of 9999, however, we don't know whether this is really the length of the longest trip or if the database was only set up to allow a four digit number.
* Let's try running the same query as before, but adjust the `LIMIT` to return the 10 highest durations and see if that's the case:

In [11]:
df=pd.read_sql_query("SELECT duration FROM trips ORDER BY duration DESC LIMIT 10",conn)
df

duration
0      9999
1      9998
2      9998
3      9997
4      9996
5      9996
6      9995
7      9995
8      9994
9      9994

### &#128681; Note:
What we see here is that there aren't a whole bunch of trips at 9999, so it doesn't look like we're cutting off the top end of our durations, but it's still difficult to tell whether that's the real length of the trip or just the maximum allowed value. Hubway charges additional fees for rides over 30 minutes (somebody keeping a bike for 9999 seconds would have to pay an extra $25 in fees) so it's plausible that they decided 4 digits would be sufficient to track the majority of rides.

# &#128640; 4. WHERE

The previous commands are great for pulling out sorted information for particular columns, but what if you already have a specific subset of the data you want to look at? That's where `WHERE` comes in. The `WHERE` command allows you to use a logical operator to specify which rows should be returned. For example you could use the following command to return every trip taken with bike `B00400`:

In [12]:
pd.read_sql_query("SELECT bike_number from trips  WHERE bike_number = 'B00400' LIMIT 10 ",conn)

bike_number
0      B00400
1      B00400
2      B00400
3      B00400
4      B00400
5      B00400
6      B00400
7      B00400
8      B00400
9      B00400

### &#128681; Note:
You'll also notice that we use quote marks in this query. That's because the bike_number is stored as a string. If the column contained numeric data types the quote marks would not be necessary.

**4.1** Let's write a query that uses WHERE to return every column in the trips table for each row with a duration longer than 9990 seconds: 

In [13]:
df=pd.read_sql_query("SELECT * FROM trips WHERE duration > 9990;",conn)
df

id  duration           start_date   ...    zip_code birth_date  gender
0     4768      9994  2011-08-03 17:16:00   ...                               
1     8448      9991  2011-08-06 13:02:00   ...                               
2    11341      9998  2011-08-09 10:42:00   ...                               
3    24455      9995  2011-08-20 12:20:00   ...                               
4    55771      9994  2011-09-14 15:44:00   ...                               
5    81191      9993  2011-10-03 11:30:00   ...                               
6    89335      9997  2011-10-09 02:30:00   ...                               
7   124500      9992  2011-11-09 09:08:00   ...                               
8   133967      9996  2011-11-19 13:48:00   ...                               
9   147451      9996  2012-03-23 14:48:00   ...                               
10  315737      9995  2012-07-03 18:28:00   ...      '02120       1964    Male
11  319597      9994  2012-07-05 11:49:00   ...                               
12  416523      9998  2012-08-15 12:11:00   ...                               
13  541247      9999  2012-09-26 18:34:00   ...                               

[14 rows x 11 columns]

### &#128681; Note: 
As you can see, the query returned 14 different trips, each with a duration of 9990 seconds or more. Something that stands out about this query is that all but one of the results has a `sub_type` of `"Casual"`. Perhaps this is an indication that `"Registered"` users are more aware of the extra fees for long trips and maybe Hubway could do a better job of conveying their pricing structure to Casual users to help them avoid overage charges.

**4.2** We can also combine multiple logical tests in our `WHERE` clause using `AND` or `OR`. If, for example, in our previous query we had only wanted to return the trips with a duration over 9990 seconds that also had a `sub_type` of `Registered`, we could use `AND` to specify both conditions. I also recommend using parentheses to separate each logical test. This isn't strictly required for the code to function, but they can make your queries easier to understand as you increase the complexity. Let's run that query now. We already know it should only return one result, so it should be easy to check that we've got it right:

In [14]:
df=pd.read_sql_query("SELECT * FROM trips WHERE (duration >= 9990) AND (sub_type = 'Registered')ORDER BY duration DESC;",conn)
df

id  duration           start_date   ...    zip_code birth_date  gender
0  315737      9995  2012-07-03 18:28:00   ...      '02120     1964.0    Male

[1 rows x 11 columns]

# &#128207; 5. COUNT
The next question we set out at the beginning of the notebook is **"How many trips were taken by `'registered'` users?" ** To answer it, we could run the same query as above and modify the `WHERE` expression to return all of the rows where `sub_type` is equal to `'Registered'` and then count them up. You could count them up on the screen, however, SQL actually has a built in command to do the work for us, `COUNT`.

This allows us to shift the calculation to the database and save us the trouble of writing additional scripts to count up results. To use it, you simply include `COUNT(column_name)` instead of (or in addition to) the columns you want to `SELECT`, like this:

In [15]:
pd.read_sql_query("SELECT COUNT(id) FROM trips ",conn)

COUNT(id)
0    1570001

### &#128681; Note: 
In this instance, it doesn't matter which column we choose to count because every column should have data for each row in our query, but sometimes a query might have missing (or "null") values for some rows. If you're not sure whether a column contains null values you can run your `COUNT` on the id column - the id column is never null, so you can be sure your count won't have missed anything. You can also use `COUNT(1)` or `COUNT(*)` to count up every row in your query.

Let's take a look at a query to answer our question. We can use `SELECT COUNT(*)` to count up the total number of columns returned and `WHERE` `sub_type` = "Registered" to make sure we only count up the trips taken by Registered users.

In [16]:
df=pd.read_sql_query("SELECT COUNT(*) FROM trips WHERE sub_type = 'Registered';",conn)
df

COUNT(*)
0   1105192

This query worked and has returned the answer to our question, but the column heading isn't particularly descriptive. If you want to make your results more readable, you can use AS to give your output an alias (or nickname). If we wanted to re-run the previous query but give our column heading an alias of Total Trips by Registered Users, we could do the following:

In [17]:
df=pd.read_sql_query(' ' ' SELECT COUNT(*) AS "Total Trips by Registered Users" FROM trips WHERE sub_type = "Registered";' ' ' ,conn)
df

Total Trips by Registered Users
0                          1105192

# &#128223; 5. Aggregate Functions(SUM,AVG,MIN,MAX)
`COUNT` is not the only trick SQL has in this domain. You can also use `SUM, AVG, MIN` and `MAX` to return the sum, average, minimum and maximum of a column respectively. These, along with `COUNT`, are known as aggregate functions. So to answer our third question,** "What was the average trip duration?"**, we can use the `AVG` function on the duration column (and, once again, use AS to give our output column a more descriptive name):

In [18]:
df=pd.read_sql_query(' ' ' SELECT AVG(duration) AS "Average Duration" FROM trips;  ' ' ' ,conn )
df

Average Duration
0        912.409682

### &#128681; Note:
So it turns out that the average trip duration is 912 seconds, which is about 15 minutes. This makes some sense since we know that Hubway charges extra fees for trips over 30 minutes because the service is designed for riders to take short, one-way trips.

But what about our next question, do registered or casual users take longer trips? We already know one way to answer this question - we can run two `SELECT AVG(duration)`` FROM` trips queries with `WHERE `clauses that restrict one to "Registered" and one to "Casual" users. SQL also includes a way that we can answer this question in a single query, using the GROUP BY command.

# &#128218; 6.GROUP BY 
`GROUP` BY separates the rows into groups based on the contents of a particular column and allows us to perform aggregate functions on each group. To get a better idea of how this works, let's take a look at the gender column. Each row can have one of three possible values in the gender column, "Male", "Female" or Null (missing - we don't have gender data for casual users). When we use GROUP BY, the database will separate out each of the rows into a different group based on the value in the gender column, in much the same way that you might separate a deck of cards into different suits. You can imagine making two piles, one of all the males, one of all the females.

So let's walk through exactly how to write a query to answer our question of whether registered or casual users take longer trips.

* As with each of our queries so far, we'll start with `SELECT` to tell the database which information we want to see. In this instance, we'll want `sub_type` and `AVG(duration)`.

* We'll also include `GROUP BY` sub_type to separate out our data by subscription type and calculate the averages of registered and casual users separately.

Here's what the code looks like when you put it all together:

In [19]:
df=pd.read_sql_query(' ' '  SELECT sub_type, AVG(duration) AS "Average Duration" FROM trips GROUP BY sub_type; ' ' ',conn)
df

sub_type  Average Duration
0      Casual       1519.643897
1  Registered        657.026067

Our next question was** which bike was used for the most trips? **and we can answer this using a very similar query. Take a look at the following example and see if you can figure out what each line is doing - we'll go through it step by step afterwards so you can check you got it right:

In [20]:
df=pd.read_sql_query(' ' ' SELECT bike_number as "Bike Number", COUNT(*) AS "Number of Trips" FROM trips GROUP BY bike_number ORDER BY COUNT(*) DESC LIMIT 1; ' ' ',conn)
df

Bike Number  Number of Trips
0      B00490             2120

### &#128681; Note:
As you can see from the output, bike B00490 took the most trips. Let's run through how we got there:

* The first line is a SELECT clause to tell the database we want to see the bike_number column and a count of every row. It also uses AS to tell the database to display each column with a more useful name.
* The second line uses FROM to specify that the data we're looking for is in the trips table.
* The third line is where things start to get a little tricky. We use GROUP BY to tell the COUNT function on line 1 to count up each value for bike_number separately.
* On line four we have an ORDER BY clause to sort the table in descending order and make sure our most-used bike is at the top.
* Finally we use LIMIT to restrict the output to the first row, which we know will be the bike that was used in the highest number of trips because of how we sorted the data on line four.

# &#128291; 7. Arithmetic Operators
The last of our questions about this table is a little more tricky than the others. We want to know the **average duration of trips by registered members over the age of 30**. We could just figure out the year in which 30 year olds were born in our heads and then plug it in, but a more elegant solution is to use arithmetic operations directly within our query. SQL allows us to use +, -, * and / to perform an arithmetic operation on an entire column at once.

In [21]:
df=pd.read_sql_query(' ' ' SELECT AVG(duration) FROM trips WHERE (2017 - birth_date) > 30; ' ' ',conn )
df

AVG(duration)
0     923.014685

# &#128279;  8. JOIN
So far we've been looking at queries that only pull data from the trips table, but you'll remember from the introduction that this database contains a second table, stations. The stations table contains information about every station in the Hubway network and includes an id column that is referenced by the trips table.
![Imgur](https://i.imgur.com/Nehnebq.png)
Before we start to work through some real examples on our dataset, let's take a moment to look back at the hypothetical order tracking database from earlier. In that database we had two tables, `orders` and `customers`, and they were connected by the `customer_id` column. Let's say we wanted to write a query that returned the `order_number` and `name` for every order in the database. If they were both stored in the same table, we could use the following query:

**```SELECT order_number, name FROM orders;```**

### &#128681; Note:
Unfortunately order_number column and the name column are stored in two different tables, so we have to add a few extra steps. 

* Let's take a moment to think through the additional things the database will need to know before it can return the information we want:

* Which table is the order_number column in?

* Which table is the name column in?

* How is the information in the orders table connected to the information in the customers table?

To answer the first two of these questions, we can include the table names for each column in our `SELECT` command. The way we do this is simply to write the table name and column name separated by a .. For example, instead of `SELECT` `order_number`, name we would write `SELECT` orders.order_number, customers.name. Adding the table names here helps the database to find the columns we're looking for by telling it which table to look in for each.

To tell the database how the orders and customers tables are connected, we use `JOIN` and `ON`. `JOIN` specifies which tables should be connected and ON specifies which columns in each table are related. We're going to use an inner join, which means that rows will only be returned where there is a match in our columns specified in `ON`. In this example, we will want to use `JOIN` on whichever table we didn't include in the `FROM` command. So we can either use `FROM` orders `INNER JOIN` customers or `FROM` customers `INNER JOIN` orders. As we discussed earlier, these tables are connected on the `customer_id` column in each table. Therefore, we will want to use `ON` to tell the database that these two columns refer to the same thing like this:

**```ON orders.customer_ID = customers.customer_id```**

Once again we use the . to make sure the database knows which table each of these columns is in. So when we put all of this together, we get a query that looks like this:

**```SELECT orders.order_number, customers.name```**

**```FROM orders```**

**```INNER JOIN customers```**

**```ON orders.customer_id = customers.customer_id```**

This query will return the order number of every order in the database along with the customer name that is associated with each.

Moving back to our Hubway database we can now write some queries to see `JOIN` in action, but before we get started we should take a look at the rest of the columns in the stations table. Here's a query to show us the first 5 rows so we can see what the stations table looks like:

In [22]:
df=pd.read_sql_query(' ' ' SELECT * FROM stations LIMIT 5; ' ' ',conn)
df

id     ...             lng
0   3     ...      -71.100812
1   4     ...      -71.069616
2   5     ...      -71.090179
3   6     ...       -71.06514
4   7     ...      -71.044624

[5 rows x 5 columns]


* `id` — A unique identifier for each station (corresponds to the start_station and end_station columns in the trips table)

* `station` — The station name

* `municipality` — The municipality that the station is in (Boston, Brookline, Cambridge or Somerville)

* `lat` — The latitude of the station

* `lng` — The longitude of the station

* Which stations are most frequently used for round trips?

* How many trips start and end in different municipalities?



I'll try to answer some questions in the data, starting with ** which station is the most frequent starting point?** Let's work through it step by step:

* First we want to use SELECT to return the station column from the stations table and the COUNT of the number of rows.
* Next we specify the tables we want to JOIN and tell the database to connect them ON the start_station column in the trips table and the id column in the stations table.
* Then we get into the meat of our query - we GROUP BY the station column in the stations table so that our COUNT will count up the number of trips for each station separately
Finally we can ORDER BY our COUNT and LIMIT the output to a manageable number of results

In [23]:
df=pd.read_sql_query(' ' '  SELECT stations.station AS "Station", COUNT(*) AS "Count" FROM trips  INNER JOIN stations ON trips.start_station = stations.id GROUP BY stations.station ORDER BY COUNT(*) DESC LIMIT 5; ' ' ',conn)
df

Station  Count
0              South Station - 700 Atlantic Ave.  56123
1       Boston Public Library - 700 Boylston St.  41994
2  Charles Circle - Charles St. at Cambridge St.  35984
3                           Beacon St / Mass Ave  35275
4                   MIT at Mass Ave / Amherst St  33644

### &#128681; Note:
If you're familiar with Boston, you'll understand why these are the most popular stations. South Station is one of the main commuter rail stations in the city, Charles Street runs along the river close to some nice scenic routes and Boylston and Beacon streets are right downtown near a number of office buildings.

The next question we'll look at is **which stations are most frequently used for round trips?** We can use much the same query as before. We will `SELECT` the same output columns and `JOIN` the tables in the same way, but this time we'll add a `WHERE` clause to restrict our `COUNT` to trips where the start_station is the same as the end_station.

In [24]:
df=pd.read_sql_query(' ' ' SELECT stations.station AS "Station", COUNT(*) AS "Count" FROM trips  INNER JOIN stations ON trips.start_station = stations.id WHERE trips.start_station = trips.end_station GROUP BY stations.station ORDER BY COUNT(*) DESC LIMIT 5; ' ' ',conn)
df

Station  Count
0    The Esplanade - Beacon St. at Arlington St.   3064
1  Charles Circle - Charles St. at Cambridge St.   2739
2       Boston Public Library - 700 Boylston St.   2548
3                  Boylston St. at Arlington St.   2163
4                           Beacon St / Mass Ave   2144

Finally,** how many trips start and end in different municipalities?** This question takes things a step further. We want to know how many trips start and end in a different municipality. To achieve this, we need to `JOIN` the trips table to the stations table twice. Once `ON` the start_station column and then `ON` the end_station column. In order to do this, we have to create an alias for the stations table so that we are able to differentiate between data that relates to the `start_station` and data that relates to the `end_station`. We do this in exactly the same way we've been creating aliases for individual columns to make them display with a more intuitive name, using `AS`.

For example we can use the following code to `JOIN` the stations table to the trips table using an alias of 'start'. We can then combine 'start' with our column names using . to refer to data that comes from this specific `JOIN` (rather than the second `JOIN` we will do `ON` the `end_station` column):

 **INNER JOIN stations AS start ON trips.start_station = start.id**

Here's what the final query will look like when we run it:

In [25]:
df=pd.read_sql_query(' ' '  SELECT COUNT(trips.id) AS "Count" FROM trips INNER JOIN stations AS start ON trips.start_station = start.id INNER JOIN stations AS end ON trips.end_station = end.id WHERE start.municipality <> end.municipality;' ' ',conn)
df

Count
0  309748

### &#128681; Note:
This shows that about 300k out of 1.5 million trips (or 20%) ended in a different municipality than they started in - further evidence that people mostly use Hubway bicycles for relatively short journeys rather than longer trips between towns.

# Motivation 
# "work HARD , dream BIG , never GIVE UP" &#128170; 